lib/internal/modules/run_main.js

```diff
@@ -162,7 +157,30 @@ function runEntryPointWithESMLoader(callback) {
 function executeUserEntryPoint(main = process.argv[1]) {
   const resolvedMain = resolveMainPath(main);
   const useESMLoader = shouldUseESMLoader(resolvedMain);
-  if (useESMLoader) {
+
+  // We might retry as ESM if the CommonJS loader throws because the file is ESM, so
+  // try the CommonJS loader first unless we know it's supposed to be parsed as ESM.
+  let retryAsESM = false;
+  if (!useESMLoader) {
+    // Module._load is the monkey-patchable CJS module loader.
+    const { Module } = require('internal/modules/cjs/loader');
+    try {
+      Module._load(main, null, true);
+    } catch (error) {
+      if (getOptionValue('--experimental-detect-module')) {
+        const { cjsRetryAsESMCache } = require('internal/modules/cjs/loader');
+        const source = cjsRetryAsESMCache.get(resolvedMain);
+        cjsRetryAsESMCache.delete(resolvedMain);
```

Comment:
I think the cache map itself can simply be released here. (Or, why do we need a map? There is only one entry point).

Reply:
My thinking was that this same approach could probably be applied to `import` of CJS modules, in which case we’d need a map? But sure, I can cross that bridge when I get to it. I also liked the extra bit of validation I get by confirming that the key of the map matches the entry filename, to ensure that we aren’t mismatching sources. If not a `SafeMap`, what do you recommend?

Reply:
The ESM loader should always have the source in some context object when it’s detecting the format. I can’t see anywhere this cache would be needed anyway….in the case of the CJS entry point a mere variable for the string is probably all it needs.

Reply:
I’m pretty sure the ESM loader determines the format in `getFormat` before any source is loaded, because it needs to determine the format in order to pick which translator to use. What I’d like to do is follow up this PR with one that takes a similar approach for `import`, so that hopefully we can avoid double parses for the case of `import` of CommonJS. I think we’ll need something like this to achieve that.

But sure, I can make it a string until a later PR needs it to be a map.